analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/LocksReleasedAllPaths.ReaderWriterLockSlim.cs

```diff
@@ -259,20 +259,140 @@ public void Throw_Finally(object param)
 
         public void IsReadLockHeld()
         {
-            readerWriterLockSlim.EnterReadLock(); // Noncompliant FP, https://github.com/SonarSource/sonar-dotnet/issues/5416
+            readerWriterLockSlim.EnterReadLock();
+            if (readerWriterLockSlim.IsReadLockHeld)    // Compliant, https://github.com/SonarSource/sonar-dotnet/issues/5416
+            {
+                readerWriterLockSlim.ExitReadLock();
+            }
+        }
+
+        public void IsReadLockHeld_NoLocking()
+        {
+            if (readerWriterLockSlim.IsReadLockHeld)    // Noncompliant
+            {
+                if (condition)
+                {
+                    readerWriterLockSlim.ExitReadLock();
+                }
+            }
+        }
+
+        public void IsReadLockHeld_NoLocking_Compliant()
+        {
             if (readerWriterLockSlim.IsReadLockHeld)
             {
                 readerWriterLockSlim.ExitReadLock();
             }
         }
 
+        public void IsReadLockHeld_Noncompliant()
+        {
+            readerWriterLockSlim.EnterReadLock();
+            if (readerWriterLockSlim.IsReadLockHeld)    // Noncompliant
+            {
+                if (condition)
+                {
+                    readerWriterLockSlim.ExitReadLock();
+                }
+            }
+        }
+
+        public void IsReadLockHeld_Noncompliant(bool arg)
+        {
+            if (arg)
+            {
+                readerWriterLockSlim.EnterReadLock();
+            }
+            if (readerWriterLockSlim.IsReadLockHeld)    // Noncompliant
+            {
+                if (condition)
+                {
+                    readerWriterLockSlim.ExitReadLock();
+                }
+            }
+        }
+
+        public void IsReadLockHeld_Unreachable()
+        {
+            readerWriterLockSlim.EnterReadLock();
+            if (readerWriterLockSlim.IsReadLockHeld)    // Noncompliant, ends up unreleased on If path, and released on Else path
+            {
+                //
+            }
+            else
+            {
+                readerWriterLockSlim.ExitReadLock();
+            }
+        }
+
         public void IsWriteLockHeld()
         {
-            readerWriterLockSlim.EnterWriteLock(); // Noncompliant FP, https://github.com/SonarSource/sonar-dotnet/issues/5416
+            readerWriterLockSlim.EnterWriteLock();  // Compliant, https://github.com/SonarSource/sonar-dotnet/issues/5416
             if (readerWriterLockSlim.IsWriteLockHeld)
             {
                 readerWriterLockSlim.ExitWriteLock();
             }
         }
+
+        public void IsWriterLockHeld_Noncompliant()
+        {
+            if (readerWriterLockSlim.IsWriteLockHeld)  // Noncompliant
```

Comment:
Please add a test case where `IsReadLockHeld` is checked and  `ExitWriteLock` is used to  release the lock. To document that we don't care about the `LockType`.